### PR TITLE
Remove all OS dependencies except for Docker

### PR DESCRIPTION
**Pull request by OrangeSquirter**

### CODE REVIEW

File: api/api/handlers.go [hunk 1]
```diff
@@ -199,11 +201,12 @@ func CreatePayloadHandler(c *gin.Context) {
 		tag := strings.Split(newPayID, "-")
 		concat := body["name"] + "_" + tag[0]
 		var commandString string
-		// Possible RCE concern
-		if body["type"] == "original" {
-			commandString = fmt.Sprintf( // Borrowed from https://github.com/s-christian/pwnts/blob/master/site/site.go#L175
 
-				"CGO_ENABLED=0 go build -trimpath -ldflags \"-s -w -X main.ServerIP=%s -X main.ServerPort=%s -X main.CallbackFrequency=%s -X main.CallbackJitter=%s -X main.RootCert=%s\" -o agents/%s client/client.go",
```
Comment (OrangeSquirter) [inline api/api/handlers.go:206]: I still don't really like this section, especially with the hardcoded paths

File: docker-compose.yml [hunk 2]
```diff
@@ -1,12 +1,14 @@
 services:
```
Comment (OrangeSquirter) [inline docker-compose.yml:1]: Ensure a bake works without running `install.sh` before merging this

File: Dockerfile.ui [hunk 3]
```diff
@@ -0,0 +1,6 @@
+FROM node:22-alpine AS build
```
Comment (OrangeSquirter) [inline Dockerfile.ui:1]: This file needs work. Want to get this working with an `npm build`

File: Dockerfile.server [hunk 4]
```diff
@@ -0,0 +1,17 @@
+FROM golang:1.22.3 AS build
+
+WORKDIR /app
+COPY . ./
+RUN go mod tidy
+RUN go build -o app ./server/server.go
+
+FROM alpine:latest
+#this seems dumb, but the libc from the build stage is not the same as the alpine libc
+#create a symlink to where it expects it since they are compatible. https://stackoverflow.com/a/35613430/3105368
```
Comment (vcaster) [inline Dockerfile.server:10]: Noice

File: api/api/handlers.go [hunk 5]
```diff
@@ -212,9 +216,10 @@ func CreatePayloadHandler(c *gin.Context) {
 				concat,
 			)
 		} else if body["type"] == "wkeys" {
-			commandString = fmt.Sprintf( // Borrowed from https://github.com/s-christian/pwnts/blob/master/site/site.go#L175
-
-				"CGO_ENABLED=0 go build -trimpath -ldflags \"-s -w -X main.ServerIP=%s -X main.ServerPort=%s -X main.CallbackFrequency=%s -X main.CallbackJitter=%s -X main.RootCert=%s\" -o agents/%s client/kclient/kclient.go",
+			commandString = fmt.Sprintf(
+				"docker run --rm -v %s:/build -w /build golang:1.22.3 "+
```
Comment (vcaster) [inline api/api/handlers.go:220]: I wonder if you can double escape docker to get to the host with this. Regardless, it's still an extra layer. Noice

File: client/kclient/kclient.go [hunk 6]
```diff
@@ -117,7 +117,7 @@ func main() {
 		}
 		ipAddress := beacon.LocalAddr().(*net.TCPAddr)
 		ip := fmt.Sprintf("%v", ipAddress)
-		init := Agentuuid + ":" + strings.TrimSuffix(string(user), "\n") + ":" + strings.TrimSuffix(string(hostname), "\n") + ":" + ip + ":NoKeysBeacon:" + ServerIP + ":" + ServerPort + ":" + CallbackFrequency + ":" + CallbackJitter + ":MASTERKEY"
+		init := Agentuuid + ":" + strings.TrimSuffix(string(user), "\n") + ":" + strings.TrimSuffix(string(hostname), "\n") + ":" + ip + ":KeysBeacon:" + ServerIP + ":" + ServerPort + ":" + CallbackFrequency + ":" + CallbackJitter + ":MASTERKEY"
```
Comment (vcaster) [inline client/kclient/kclient.go:120]: This should be NoKeysBeacon still. Line 222 covers this

File: docker-bake.hcl
Comment (vcaster) [inline docker-bake.hcl:1]: Ouuuu fancy

File: docker-compose.yml [hunk 7]
```diff
@@ -1,53 +1,79 @@
...
-    command: sh -c "go run server/server.go"
-    working_dir: /app
     volumes:
-      - ./:/app
+      - ./logs:/app/logs
+      - ./certs:/app/certs
     networks:
-      internal_c2_network:
-        ipv4_address: "${DOCKER_INTERNAL}.5"
+      - internal_c2_network
```
Comment (vcaster) [inline docker-compose.yml:20]: irc the last octet was defined on purpose. Did you test this out?

File: install.sh [hunk 8]
```diff
@@ -1,50 +1,54 @@
...
 function ask_prompt {
-   
-   echo "Note: Webserver, ReactClient, DB and C2 server can't be on the same port and must be an unused port"
-   read -p "Enter APISERVER IP: " webserverip
+   echo "Note: Webserver, ReactClient, DB, and C2 server can't be on the same port and must be an unused port"
    read -p "Enter APISERVER PORT: " webserverport
    read -p "Enter REACTCLIENT IP: " reactclientip
    read -p "Enter REACTCLIENT PORT: " reactclientport
-   echo "Note: To listen on all inteface, leave C2SERVER IP blank"
+   echo "Note: To listen on all interfaces, leave C2SERVER IP blank"
```
Comment (vcaster) [inline install.sh:48]: Thanks lol

File: install.sh [hunk 9]
```diff
@@ -152,116 +162,67 @@ if [ -z "$default" ]; then
...
-go mod tidy
-
-# npm install
-echo "Installing node modules..."
-
-cd ui && npm install && cd ../
-
-echo ""
-echo ""
-echo "------------------------------------------Raw Dog Run------------------------------------------"
```
Comment (vcaster) [inline install.sh:240]: RIP, Raw Dog Run

**Review by vcaster — APPROVED** [review]
Approved with comments

File: client/kclient/kclient.go (hunk 6 above)
Comment (OrangeSquirter) [reply client/kclient/kclient.go:120]: ah ok, it confused me when seeing it in the logs

File: docker-compose.yml (hunk 7 above)
Comment (OrangeSquirter) [reply docker-compose.yml:20]: yes, it is using the internal Docker DNS. Tested on 3 separate boxes and all of them functioned fully

File: api/api/handlers.go (hunk 5 above)
Comment (OrangeSquirter) [reply api/api/handlers.go:220]: could prob break out into the build container, which only lasts a few seconds. With the recent UI changes with auth, im not too worried

File: install.sh (hunk 9 above)
Comment (OrangeSquirter) [reply install.sh:240]: Being on the docker network, raw dog runs haven't worked in a hot minute. With some manipulation of the docker-compose file, can still do it for the server and UI

File: install.sh (hunk 8 above)
Comment (OrangeSquirter) [reply install.sh:45]: ChatGPT spell check lol

**Pull request merged**